### PR TITLE
Add Support for  BAAI/bge-m3

### LINKED ISSUE #107: Support BAAI/bge-m3
Requesting support for BAAI/bge-m3. Thanks.

### PR DESCRIPTION
**Pull request by Ya-shh**
Canonical vector values: https://colab.research.google.com/drive/1OTZcLxWWAmJqV1ZDS57W5x6zJPRzku4n?usp=sharing

Onnx model: https://huggingface.co/yashvardhan7/bge-m3-onnx/tree/main

Contributed  ONNX weights to the BAAI/bge-m3 HF model repo :https://huggingface.co/BAAI/bge-m3/tree/main/onnx

### CODE REVIEW

**Comment by Ya-shh** [comment]
@Anush008 this one has passed all the tests locally .Additionally I have updated `onnx_embedding.py` file as it was generating this error :
test_embedding():
        is_ubuntu_ci = os.getenv("IS_UBUNTU_CI")
    
        for model_desc in TextEmbedding.list_supported_models():
            if is_ubuntu_ci == "false" and model_desc["size_in_GB"] > 1:
                continue
    
            dim = model_desc["dim"]
            model = TextEmbedding(model_name=model_desc["model"])
    
            docs = ["hello world", "flag embedding"]
>           embeddings = list(model.embed(docs))

fastembed/tests/test_text_onnx_embeddings.py:48: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
fastembed/fastembed/text/text_embedding.py:89: in embed
    yield from self.model.embed(documents, batch_size, parallel, **kwargs)
fastembed/fastembed/text/onnx_embedding.py:223: in embed
    yield from self._embed_documents(
fastembed/fastembed/common/onnx_model.py:97: in _embed_documents
    yield from self._post_process_onnx_output(self.onnx_embed(batch))
fastembed/fastembed/common/onnx_model.py:70: in onnx_embed
    model_output = self.model.run(None, onnx_input)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <onnxruntime.capi.onnxruntime_inference_collection.InferenceSession object at 0x10217eb90>
output_names = ['token_embeddings', 'sentence_embedding']
input_feed = {'attention_mask': array([[1, 1, 1, 1, 1, 0],
       [1, 1, 1, 1, 1, 1]]), 'input_ids': array([[    0, 33600,    31,  ...[    0, 49938,     6, 55720, 59725,     2]]), 'token_type_ids': array([[0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0]])}
run_options = None

    def run(self, output_names, input_feed, run_options=None):
        """
        Compute the predictions.
    
        :param output_names: name of the outputs
        :param input_feed: dictionary ``{ input_name: input_value }``
        :param run_options: See :class:`onnxruntime.RunOptions`.
        :return: list of results, every result is either a numpy array,
            a sparse tensor, a list or a dictionary.
    
        ::
    
            sess.run([output_name], {input_name: x})
        """
        self._validate_input(list(input_feed.keys()))
        if not output_names:
            output_names = [output.name for output in self._outputs_meta]
        try:
>           return self._sess.run(output_names, input_feed, run_options)
E           onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Invalid input name: token_type_ids

../.pyenv/versions/3.10.10/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py:220: InvalidArgument

**Comment by Ya-shh** [comment]
Fixes:#107

File: fastembed/text/onnx_embedding.py [hunk 1]
```diff
@@ -217,11 +227,29 @@ def embed(
...
+        }
 
+        # Check if 'token_type_ids' is expected by the model
+        if 'token_type_ids' in [input.name for input in self.model.get_inputs()]:
+            token_type_ids = np.zeros_like(input_ids)
+            onnx_input["token_type_ids"] = token_type_ids.astype(np.int64)
+
+        model_output = self.model.run(None, onnx_input)
+        embeddings = model_output[0]
+        return embeddings, attention_mask
```
Comment (Anush008) [inline fastembed/text/onnx_embedding.py:248]: We can override `_preprocess_onnx_input()` instead.

For example https://github.com/qdrant/fastembed/blob/main/fastembed%2Ftext%2Fe5_onnx_embedding.py#L45-L50

Comment (Ya-shh) [reply fastembed/text/onnx_embedding.py:248]: @Anush008 ,Initially I did that only but the error  persists then only  I modified [onnx_embedding.py ](https://github.com/qdrant/fastembed/pull/197/files#diff-c3531be71315167c13fae0ab3afeb93d52ec56bf9f004e4a60675e369530c4e8)

Comment (Ya-shh) [reply fastembed/text/onnx_embedding.py:248]: You could also try it locally

Comment (Ya-shh) [reply fastembed/text/onnx_embedding.py:248]: @Anush008 ,It won't work as `e5_onnx_embedding.py` is  different. The model is being stored in `onnx_embedding.py`. It will work for an e5 model. The same error occurred while uploading  `e5-small` and it resolved  by modifying the `  def _preprocess_onnx_input` . But for `bge-m3` it isn't working.

Comment (Ya-shh) [reply fastembed/text/onnx_embedding.py:248]: for `multilingual-e5-small model` , I override like this and it worked :  

    def _preprocess_onnx_input(self, onnx_input: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
        """
        Preprocess the onnx input.
        """
        if "token_type_ids" not in self.required_inputs:
            onnx_input.pop("token_type_ids", None)
        return onnx_input

But it failed for `bge-m3`

Comment (Ya-shh) [reply fastembed/text/onnx_embedding.py:248]: https://github.com/qdrant/fastembed/pull/190/files#diff-d029a94e189b08dcd60a4efc05e94ce6af8f51ebbda58eecdc6413e0d439eb14

Comment (Ya-shh) [reply fastembed/text/onnx_embedding.py:248]: [CI logs](https://github.com/qdrant/fastembed/actions/runs/8626874505/job/23645811364#step:5:21)

**Comment by Ya-shh** [comment]
@Anush008 any updates from your side ?

File: fastembed/text/onnx_embedding.py (hunk 1 above)
Comment (Anush008) [reply fastembed/text/onnx_embedding.py:248]: You're saying overriding `_preprocess_onnx_input()` in `onnx_embedding.py` doesn't work?

Comment (Ya-shh) [reply fastembed/text/onnx_embedding.py]: I tried it another way ,now it's working .All tests have passed

**Comment by Anush008** [comment]
CI passed. Awesome.

I see you've contributed the ONNX weights to the BAAI/bge-m3 HF model repo.

https://huggingface.co/BAAI/bge-m3

Would it make sense to use it as the source?

**Comment by Ya-shh** [comment]
> CI passed. Awesome.
> 
> I see you've contributed the ONNX weights to the BAAI/bge-m3 HF model repo.
> 
> https://huggingface.co/BAAI/bge-m3
> 
> We can now use the original repo as the source?

Done !

**Review by Anush008 — APPROVED** [review]
🚀🚀

**Comment by Ya-shh** [comment]
@NirantK If I get to know the steps, I can definitely do this. However, I haven't tried this before. Could you please direct me to the initial steps for executing this task?

**Comment by Ya-shh** [comment]
@NirantK  , I found a `bge-m3` model on [HF](https://huggingface.co/ddmitov/bge_m3_dense_colbert_onnx) which supports both dense vectors and ColBERt. I even tested the outputs on colab :https://colab.research.google.com/drive/1-ZCraKfL7cLpFFD8tKmo1k7dEof1QFWf?usp=sharing.

**Comment by Ya-shh** [comment]
however , when I tested with my bge-m3 onnx model repo .This was the output:https://colab.research.google.com/drive/1daDierjr5bvgk5aWrhJ6adSqMF1H1R1e?usp=sharing

**Comment by Ya-shh** [comment]
Could we test `bge-m3-onnx`  with SPLADE vectors  like this ? Notebook:https://colab.research.google.com/drive/1phjnxXzUzOMd_x3_015ZImjnBuio5IbW?usp=sharing

**Comment by Ya-shh** [comment]
@NirantK could you please review these so that we can decide which model we should upload?

**Comment by Ya-shh** [comment]
@Anush008 could you  please review these models ?

**Comment by Anush008** [comment]
We'll have to wait for Nirant to take a look at those.

**Comment by Ya-shh** [comment]
@NirantK could you  please review this ?

**Comment by NirantK** [comment]
First off, thanks a ton for the love and contributions @Ya-shh  — really appreciate. 

Thank you for the investigations and looking around for more ONNX exports e.g. [aapot](https://huggingface.co/aapot/bge-m3-onnx), [ddmitov](https://huggingface.co/ddmitov/bge_m3_dense_colbert_onnx), official [BGE-M3](https://huggingface.co/BAAI/bge-m3/tree/main/onnx) — none of them do all 3 as far as I can tell 

Given the time and effort you've put across PRs, I believe it'd be helpful to expand how I'm thinking about this right now: 

## BGE-M3

1. BGE-M3 should exist in a separate, new category e.g. `multi` like we do for `sparse`
2. It should support all 3 — not just dense and ColBERT. This means FastEmbed will have to store and do some of the normalisation, loadings heads (for ColBERT and SPLADE) and passing attention which FlagEmbedding is doing right now. This PR should include those changes. 

## Tests and Notebooks
 
The notebooks are a bit difficult to make sense of: 

1. [SPLADE notebook](https://colab.research.google.com/drive/1phjnxXzUzOMd_x3_015ZImjnBuio5IbW?usp=sharing) has PyTorch outputs, not ONNX. We'd need both to compare and use tests for both SPLADE, Dense and write new ones for ColBERT

2. ColBERT notebooks: I don't see a comparison or test in both of them? But the dimensions looks different and I assume that's because of different normalisation or export mismatch between the two models

## Suggestions

1. Perhaps contribute a simple ColBERT notebook in a new category first? That will allow us to collaborate on ColBERT tests as well. I suspect this is the most popular model and a great starting point: https://github.com/stanford-futuredata/ColBERT 

2. Please look into closing existing [open PRs](https://github.com/qdrant/fastembed/pull/181) before opening new ones? It's not a great idea to have lot of open streams of work. If the ONNX export needs more investigation. I would love it if you took that up and contributed the solution here as an example notebook, I'll be happy and excited to review that!

**Comment by Ya-shh** [comment]
@NirantK Thank you immensely for your kind words and recognition. Your appreciation means a lot to me!
Your detailed feedback is incredibly helpful in guiding my efforts further. I'm deeply grateful for the opportunity to expand on my contributions and improve the ONNX exports. I wholeheartedly agree with your suggestions and will prioritize them accordingly.

⎯Regarding your thoughts on `BGE-M3`, creating a separate category `multi` , for that I will definitely  implement the necessary changes to support all three aspects efficiently.
As for the notebooks, I understand the importance of clarity and consistency. Rest assured, I'll work diligently to ensure they provide comprehensive comparisons and tests for all three vectors.
Your suggestion to begin with a `ColBERT` notebook is well-taken, and I'll proceed with enthusiasm. Collaborating on tests for this popular model is a fantastic opportunity to enhance our collective efforts.

Lastly, I appreciate your reminder to close existing PRs before opening new ones. Streamlining our workflow will undoubtedly lead to greater efficiency and effectiveness. I'll make it a priority to address any outstanding tasks promptly.

Once again, thank you for your unwavering support and encouragement. I'm too excited to delve deeper into the ONNX export investigation and share my findings with you through an example notebook. Your review and feedback are eagerly anticipated!

**Comment by zolero** [comment]
Any updates on this?
